File: app/src/main/java/de/christian2003/smarthome/view/url/UrlViewModel.java
package de.christian2003.smarthome.view.url;

import android.util.Patterns;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModel;
import java.net.MalformedURLException;
import java.net.URL;
import de.christian2003.smarthome.model.Config;


/**
 * Class implements the view model for the {@link UrlActivity}.
 */
public class UrlViewModel extends ViewModel {

    /**
     * Attribute stores the URL entered by the user. This is {@code null} if no URL was entered.
     */
    @Nullable
    private URL url;


    /**
     * Constructor instantiates a new view model.
     */
    public UrlViewModel() {
        url = null;
        updateUrl(Config.getInstance().getServerUrl(null));
    }


    /**
     * Method returns the URL string entered by the user.
     *
     * @return  URL string entered by the user.
     */
    @Nullable
    public String getUrl() {
        if (url == null) {
            return null;
        }
        else {
            return url.toString();
        }
    }


    /**
     * Method changes the URL entered by the user to the passed char sequence. The URL within the
     * view model is only updated if the passed argument is valid (i.e. the method returns
     * {@code true}).
     *
     * @param url   New URL entered by the user.
     * @return      Whether the entered URL is valid.
     */
    public boolean updateUrl(@Nullable CharSequence url) {
        if (url == null) {
            return false;
        }
        if (Patterns.WEB_URL.matcher(url).matches()) {
            try {
                this.url = new URL(url.toString());
            }
            catch (MalformedURLException e) {
                return false;
            }
            return true;
        }
        return false;
    }


    /**
     * Method saves the URL entered by the user to the storage.
     */
    public void saveUrl() {
        if (url != null) {
            String urlString = url.toString();
            Config.getInstance().setServerUrl(urlString);
        }
    }

}

File: common/data/src/main/java/de/christian2003/smarthome/data/view/url/UrlViewModel.kt
package de.christian2003.smarthome.data.view.url

import android.app.Application
import android.content.Context
import android.util.Patterns
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.AndroidViewModel
import java.net.URL


/**
 * Class implements the view model for the view through which to edit the URL.
 */
class UrlViewModel(application: Application): AndroidViewModel(application) {

    /**
     * Attribute stores the preferences in which to store the URL.
     */
    private val preferences = getApplication<Application>().getSharedPreferences("smart_home", Context.MODE_PRIVATE)

    /**
     * Attribute stores the URL to edit.
     */
    var url: String by mutableStateOf("")

    /**
     * Attribute indicates whether the URL is valid.
     */
    var isUrlValid: Boolean by mutableStateOf(true)


    /**
     * Method instantiates the view model.
     */
    fun init() {
        url = preferences.getString("server_url", "")!!
    }


    /**
     * Method determines whether the URL entered is valid.
     */
    fun isUrlValid() {
        if (Patterns.WEB_URL.matcher(url).matches()) {
            try {
                URL(this.url)
                isUrlValid = true
            }
            catch (e: Exception) {
                isUrlValid = false
            }
        }
        else {
            isUrlValid = false
        }
    }


    /**
     * Method saves the URL entered.
     *
     * @return  Whether the URL entered was saved successfully.
     */
    fun saveUrl(): Boolean {
        if (isUrlValid) {
            val editor = preferences.edit()
            editor.putString("server_url", url)
            editor.apply()
            return true
        }
        else {
            return false
        }
    }

}
